autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/securestorage/SecureStorageDatabaseFactory.kt

```diff
@@ -18,28 +18,55 @@ package com.duckduckgo.autofill.impl.securestorage
 
 import android.content.Context
 import androidx.room.Room
+import com.duckduckgo.autofill.api.AutofillFeature
 import com.duckduckgo.di.scopes.AppScope
 import com.duckduckgo.securestorage.store.db.ALL_MIGRATIONS
 import com.duckduckgo.securestorage.store.db.SecureStorageDatabase
 import com.squareup.anvil.annotations.ContributesBinding
 import dagger.SingleInstanceIn
 import javax.inject.Inject
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
 import net.sqlcipher.database.SupportFactory
 
 interface SecureStorageDatabaseFactory {
-    fun getDatabase(): SecureStorageDatabase?
+    suspend fun getDatabase(): SecureStorageDatabase?
 }
 
 @SingleInstanceIn(AppScope::class)
 @ContributesBinding(AppScope::class)
 class RealSecureStorageDatabaseFactory @Inject constructor(
     private val context: Context,
     private val keyProvider: SecureStorageKeyProvider,
+    private val autofillFeature: AutofillFeature,
 ) : SecureStorageDatabaseFactory {
     private var _database: SecureStorageDatabase? = null
 
+    private val mutex = Mutex()
+
+    override suspend fun getDatabase(): SecureStorageDatabase? {
+        return if (autofillFeature.canCreateAsyncPreferences().isEnabled()) {
+            getAsyncDatabase()
+        } else {
+            getDatabaseSynchronized()
+        }
+    }
+
     @Synchronized
-    override fun getDatabase(): SecureStorageDatabase? {
+    private fun getDatabaseSynchronized(): SecureStorageDatabase? {
+        return runBlocking {
+            getInnerDatabase()
+        }
+    }
+
+    private suspend fun getAsyncDatabase(): SecureStorageDatabase? {
+        return mutex.withLock {
+            getInnerDatabase()
+        }
+    }
```

Comment:
I think we should do double-check here
```suggestion
    private suspend fun getAsyncDatabase(): SecureStorageDatabase? {
        _database?.let { return it }
        return mutex.withLock {
            getInnerDatabase()
        }
    }
```

Reply:
Ah yes, I missed it when removing duplicate code. Thanks!